utils-macro/src/lib.rs

```diff
@@ -0,0 +1,27 @@
+use proc_macro::TokenStream;
+use quote::ToTokens;
+use syn::{parse_macro_input, parse_quote, Block, ItemFn};
+
+#[proc_macro_attribute]
+pub fn enable_logging(
+    _attr: TokenStream,
+    item: TokenStream,
+) -> TokenStream {
+    let mut function = parse_macro_input!(item as ItemFn);
+
+    let name = function.sig.ident.to_string();
+    let stmts = function.block.stmts;
+    let block: Block = parse_quote! {{
+        if ::std::env::args().any(|e| e == "--nocapture") {
+            let subscriber = ::tracing::subscriber::set_default(::miden_node_utils::logging::subscriber());
+            let span = ::tracing::span!(::tracing::Level::INFO, "Test run", name = #name);
+
+            #(#stmts)*
```

Comment:
I couldn't find a way to emit logs only if a test fail with the default test runner built in `cargo test` (and I also have not searched for an alternative test runner). So this will be quite noisy, one thing that we could implement is a [log line grouping](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines), but I think this is not super necessary for now.

Reply:
One potential option is to use [tracing-forest](https://docs.rs/tracing-forest/latest/tracing_forest/). It is supposed to display related spans in an easy to visualize way. We are using it Miden VM - but not across async boundaries.

Reply:
`tracing-forest` seems like an improvement, but doesn't fix the issue I was talking about. I added it as a layer to test it out, here is what I got for `cargo test -- test_build_batch --nocapture`:

```
2024-01-31T13:06:42.780217Z  INFO Test run: miden_node_block_producer::txqueue::tests: block-producer/src/txqueue/tests/mod.rs:75: new name: "test_build_batch_success"
2024-01-31T13:06:42.780218Z  INFO Test run: miden_node_block_producer::txqueue::tests: block-producer/src/txqueue/tests/mod.rs:148: new name: "test_build_batch_failure"
2024-01-31T13:06:42.781495Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:193: new
2024-01-31T13:06:42.781495Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:193: new
2024-01-31T13:06:42.781552Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:198: tx_id: 0x89eaa76646fd783a2a9f59fd497c8006302b980382a9be8885d116ae30148149, account_id: 0x2cf0094fbccda1df
2024-01-31T13:06:42.781556Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:198: tx_id: 0x89eaa76646fd783a2a9f59fd497c8006302b980382a9be8885d116ae30148149, account_id: 0x2cf0094fbccda1df
2024-01-31T13:06:42.781591Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:211: Transaction added to tx queue, queue_len: 1
2024-01-31T13:06:42.781591Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:211: Transaction added to tx queue, queue_len: 1
2024-01-31T13:06:42.781620Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:193: close, time.busy: 82.2µs, time.idle: 44.0µs
2024-01-31T13:06:42.781623Z  INFO add_transaction: miden-block-producer: block-producer/src/txqueue/mod.rs:193: close, time.busy: 85.1µs, time.idle: 44.5µs
de4aa47e-da94-4a2e-a5f4-4e5345509631 2024-01-31T13:06:42.781517+00:00     INFO     add_transaction [ 81.2µs | 100.00% ]
de4aa47e-da94-4a2e-a5f4-4e5345509631 2024-01-31T13:06:42.781575+00:00     INFO     ┝━ ｉ [info]:  | tx_id: 0x89eaa76646fd783a2a9f59fd497c8006302b980382a9be8885d116ae30148149 | account_id: 0x2cf0094fbccda1df
de4aa47e-da94-4a2e-a5f4-4e5345509631 2024-01-31T13:06:42.781604+00:00     INFO     ┕━ ｉ [info]: Transaction added to tx queue | queue_len: 1
5da4dcd6-3a24-4e5b-a4d6-e1b72dbc3d54 2024-01-31T13:06:42.781526+00:00     INFO     add_transaction [ 84.8µs | 100.00% ]
5da4dcd6-3a24-4e5b-a4d6-e1b72dbc3d54 2024-01-31T13:06:42.781570+00:00     INFO     ┝━ ｉ [info]:  | tx_id: 0x89eaa76646fd783a2a9f59fd497c8006302b980382a9be8885d116ae30148149 | account_id: 0x2cf0094fbccda1df
5da4dcd6-3a24-4e5b-a4d6-e1b72dbc3d54 2024-01-31T13:06:42.781610+00:00     INFO     ┕━ ｉ [info]: Transaction added to tx queue | queue_len: 1
```

We have duplicated lines, at the top are the entries from the `tracing_subscriber`'s pretty printer, followed by the lines from the `tracing-forest`. To me it seems that tracing-forest is clearly better in this context, since the lines for the `add_trasaction` span are correlated and grouped together.

But what I meant, is that the line for the new test, added [here](https://github.com/0xPolygonMiden/miden-node/pull/199/files#diff-0aaf9402448e671940f73ab4211e21e597f07f1bedcaea42545ebf6209005ddbR17) is not being carried over to the test execution. We have the opening line at the top, printed by the `tracing_subscriber`, and then at the bottom of the test we have:

```
2024-01-31T13:06:42.795733Z  INFO Test run: miden_node_block_producer::txqueue::tests: block-producer/src/txqueue/tests/mod.rs:75: close, time.busy: 0.00ns, time.idle: 15.5ms name: "test_build_batch_success"
f5470aac-853a-408b-b8b3-f7c8ffccb541 2024-01-31T13:06:42.780249+00:00     INFO     Test run [ 0.00ns | NaN% ] name: "test_build_batch_success"
```

Which are the closing lines for the whole test. For some reason this span didn't become the parent of all other spans in the test. I didn't spend time trying to figure out why though.

Reply:
here is the compiler's output after macros are expanded for [`test_build_block_failure`](https://github.com/0xPolygonMiden/miden-node/blob/21d85ec7726411c201020f8e34205c2e1883cd39/block-producer/src/block_builder/tests.rs#L70-L82):

```rust
        extern crate test;
        #[cfg(test)]
        #[rustc_test_marker = "block_builder::tests::test_build_block_failure"]
        pub const test_build_block_failure: test::TestDescAndFn = test::TestDescAndFn {
            desc: test::TestDesc {
                name: test::StaticTestName(
                    "block_builder::tests::test_build_block_failure",
                ),
                ignore: false,
                ignore_message: ::core::option::Option::None,
                source_file: "block-producer/src/block_builder/tests.rs",
                start_line: 73usize,
                start_col: 10usize,
                end_line: 73usize,
                end_col: 34usize,
                compile_fail: false,
                no_run: false,
                should_panic: test::ShouldPanic::No,
                test_type: test::TestType::UnitTest,
            },
            testfn: test::StaticTestFn(|| test::assert_test_result(
                test_build_block_failure(),
            )),
        };
        /// Tests that `build_block()` fails when `get_block_inputs()` fails
        fn test_build_block_failure() {
            if ::std::env::args().any(|e| e == "--nocapture") {
                let subscriber = ::tracing::subscriber::set_default(
                    ::miden_node_utils::logging::subscriber(),
                );
                let span = {
                    use ::tracing::__macro_support::Callsite as _;
                    static __CALLSITE: ::tracing::callsite::DefaultCallsite = {
                        static META: ::tracing::Metadata<'static> = {
                            ::tracing_core::metadata::Metadata::new(
                                "Test run",
                                "miden_node_block_producer::block_builder::tests",
                                ::tracing::Level::INFO,
                                ::core::option::Option::Some(
                                    "block-producer/src/block_builder/tests.rs",
                                ),
                                ::core::option::Option::Some(72u32),
                                ::core::option::Option::Some(
                                    "miden_node_block_producer::block_builder::tests",
                                ),
                                ::tracing_core::field::FieldSet::new(
                                    &["name"],
                                    ::tracing_core::callsite::Identifier(&__CALLSITE),
                                ),
                                ::tracing::metadata::Kind::SPAN,
                            )
                        };
                        ::tracing::callsite::DefaultCallsite::new(&META)
                    };
                    let mut interest = ::tracing::subscriber::Interest::never();
                    if ::tracing::Level::INFO
                        <= ::tracing::level_filters::STATIC_MAX_LEVEL
                        && ::tracing::Level::INFO
                            <= ::tracing::level_filters::LevelFilter::current()
                        && {
                            interest = __CALLSITE.interest();
                            !interest.is_never()
                        }
                        && ::tracing::__macro_support::__is_enabled(
                            __CALLSITE.metadata(),
                            interest,
                        )
                    {
                        let meta = __CALLSITE.metadata();
                        ::tracing::Span::new(
                            meta,
                            &{
                                #[allow(unused_imports)]
                                use ::tracing::field::{debug, display, Value};
                                let mut iter = meta.fields().iter();
                                meta.fields()
                                    .value_set(
                                        &[
                                            (
                                                &::core::iter::Iterator::next(&mut iter)
                                                    .expect("FieldSet corrupted (this is a bug)"),
                                                ::core::option::Option::Some(
                                                    &"test_build_block_failure" as &dyn Value,
                                                ),
                                            ),
                                        ],
                                    )
                            },
                        )
                    } else {
                        let span = ::tracing::__macro_support::__disabled_span(
                            __CALLSITE.metadata(),
                        );
                        {};
                        span
                    }
                };
                let body = async {
                    let store = Arc::new(MockStoreFailure);
                    let block_builder = DefaultBlockBuilder::new(
                        store.clone(),
                        store.clone(),
                    );
                    let result = block_builder.build_block(&Vec::new()).await;
                    if !match result {
                        Err(BuildBlockError::GetBlockInputsFailed(_)) => true,
                        _ => false,
                    } {
                        ::core::panicking::panic(
                            "assertion failed: matches!(result, Err(BuildBlockError :: GetBlockInputsFailed(_)))",
                        )
                    }
                };
                let mut body = body;
                #[allow(unused_mut)]
                let mut body = unsafe {
                    ::tokio::macros::support::Pin::new_unchecked(&mut body)
                };
                let body: ::core::pin::Pin<
                    &mut dyn ::core::future::Future<Output = ()>,
                > = body;
                #[allow(clippy::expect_used, clippy::diverging_sub_expression)]
                {
                    return tokio::runtime::Builder::new_current_thread()
                        .enable_all()
                        .build()
                        .expect("Failed building the Runtime")
                        .block_on(body);
                }
            } else {
                let body = async {
                    let store = Arc::new(MockStoreFailure);
                    let block_builder = DefaultBlockBuilder::new(
                        store.clone(),
                        store.clone(),
                    );
                    let result = block_builder.build_block(&Vec::new()).await;
                    if !match result {
                        Err(BuildBlockError::GetBlockInputsFailed(_)) => true,
                        _ => false,
                    } {
                        ::core::panicking::panic(
                            "assertion failed: matches!(result, Err(BuildBlockError :: GetBlockInputsFailed(_)))",
                        )
                    }
                };
                let mut body = body;
                #[allow(unused_mut)]
                let mut body = unsafe {
                    ::tokio::macros::support::Pin::new_unchecked(&mut body)
                };
                let body: ::core::pin::Pin<
                    &mut dyn ::core::future::Future<Output = ()>,
                > = body;
                #[allow(clippy::expect_used, clippy::diverging_sub_expression)]
                {
                    return tokio::runtime::Builder::new_current_thread()
                        .enable_all()
                        .build()
                        .expect("Failed building the Runtime")
                        .block_on(body);
                }
            };
        }
```

so the macro expansions is happening from top-to-bottom. Meaning the tokio runtime is expanded first, and then the logging configuration. So this does indeed cross from sync to async land

Reply:
Okay, I was looking at the wrong problem. Turns out I just missed the manual call to enter the span 😬. Fixing it.